### PR TITLE
[EXTENDEDSETTINGS] Implement DispCal: Display White point Calibration

### PR DESCRIPTION
**Pull request by kholk**
Recently I wrote a PCC Profiling facility in the somc_panel
    kernel driver for SoMC devices displays.
    The PCC profiling facility allows a power user, or a system
    application to apply various predefined calibrations to the
    device's display.
    
    This implements the usage of this new kernel functionality
    for the end user through an easy menu to choose between
    all the available predefined whitepoint calibrations.

    Save the user-chosen white point calibration to a system
    property (persist.dispcal.setting) and implement a receiver
    that listens to the boot intents, such as PRE_BOOT_COMPLETED,
    LOCKED_BOOT_COMPLETED and BOOT_COMPLETED (for pre-N compat)
    and restores the user setting at boot time.

### CODE REVIEW

**Comment by djselbeck** [comment]
My kagura freezes on activation of the BootBroadcastReceiver . When I only apply commit ac55b84 it is working but of course does not apply it after reboot.

**Comment by djselbeck** [comment]
It does not relate to the receiver but to the kernel itself. When I write a 2 to /sys/devices/mdss_dsi_panel/pcc_profile from an adb root shell it also freezes. 

Strange because I can change the profile from the ExtendedSettings when I disable the BootBroadcastReceiver

**Comment by kholk** [comment]
Interesting, definitely. Will check, thanks!

**Comment by djselbeck** [comment]
This only happens if the device is not yet completely booted up. For example during the encryption prompt. As soon as I can enter my sim pin the kernel sysfs file works as expected (if this helps), 

btw. thanks for implementing this feature, really like the ability to set the display to a warmer white :)

**Comment by kholk** [comment]
Can you give me some more details, please?

1. Are you on SELinux enforcing?
2. Do a "dmesg | grep Panel" (quotes excluded, obviously), paste here the string that says "Panel Name: (name)"
3. Which ROM are you using right now?
4. Are you on sonyxperiadev/kernel HEAD?


Asking for this because another guy says he has no issues on Kagura.

Thanks!

**Comment by djselbeck** [comment]
Of course,

1. permissive 
2. 

> kagura:/ # dmesg|grep Panel                                                                                                                                                                                       
[    2.411629] do_panel_detect: WARNING: Panel detection LCD ID setup failed.
[    2.411761] do_panel_detect: Starting ADC Panel Detection...
[    2.412156] mdss_panel_parse_dt: Panel Name = 6


3. self-compiled 

> [ro.build.date.utc]: [1507901550]
[ro.build.description]: [aosp_f8331-userdebug 8.0.0 OPR3.170623.008 eng.hb.20171013.153230 test-keys]
[ro.build.display.id]: [aosp_f8331-userdebug 8.0.0 OPR3.170623.008 eng.hb.20171013.153230 test-keys]
[ro.build.fingerprint]: [Sony/aosp_f8331/kagura:8.0.0/OPR3.170623.008/hb10131532:userdebug/test-keys]


4. Yes

I also don't understand it as it works perfectly after device startup. I wonder if it has something to do with date set to 1970 until boot up? But I fail to imagine to see a connection between these two things

**Comment by kholk** [comment]
That's even more interesting. The other guy has the same panel as you.
I imagine "self-compiled" also means "AOSP", so we can factor out any custom ROM mess (even though it shouldn't really matter for this issue).... and being on kernel HEAD makes it even more complicated, as your setup has only one difference from the one from the other guy.

I don't think that the feature hates being used prior to the date it got created, so that clock thing can be also safely factored out.
...seriously, no that can't be a problem.

The only thing I can think about now is the encryption..... have you got any pstore log from that freeze?
Then, does it freeze-reboot (kernel panic, blinking red led at reboot) or does it freeze like HW fault?

Another thing, just to be completely paranoid, can you try booting with your SIM off of it?
There are two things I suspect:
- SIM PIN makes security to go nuts for some reason
- Encryption makes security to go nu... etc

Basically nothing else can go wrong.
It is crucial for me to reproduce the issue locally, so that I can perhaps try to add some mutex locking to the kernel function that is responsible for updating the MDP HW.

**Comment by djselbeck** [comment]
Thanks a lot for your effort.

Yes no messing around with some BS custom rom ;).

So I have one more observation, if I run the command

> adb  root && sleep 1 && adb shell "echo 2 > /sys/devices/mdss_dsi_panel/pcc_profile"

directly after the devices switches from the white Sony logo to the black android animation, it will successfully apply the correct profile.

If I wait until the prompt of the encryption PW and then run the above command it will freeze.
And if I wait until the second animation (after entering the encryption PW) it will also apply it correctly

**Comment by kholk** [comment]
So basically what craps the thing out is some sort of security that the encrypted device password stage is applying to the system.
Then I have to check if there's any way to avoid touching the sysfs while the device is in that stage!

**Comment by kholk** [comment]
Okay! I've had a fast conversation with @oshmoun. He pointed out that the Android properties are getting read in /data..... and what's encrypted is... /data.

So... now you know why I'm stupid :D

Thanks for reporting, will fix asap!

**Comment by djselbeck** [comment]
Thanks. But why would the panel driver use /data?

**Comment by kholk** [comment]
It... Doesn't...
Yeah that is still something we didn't explain to ourselves, but the app requires to read the property, which is stored in /data, so for now I will work around it by delaying it after the encryption pw screen.

Whenever I'll have time to dig, I will ;)

**Comment by djselbeck** [comment]
I think in some states something of the panel driver does not like the reset and resetup. 

Maybe it has something to do with refreshing of the screen? Hence its working during the animation at boot?

**Comment by kholk** [comment]
@djselbeck sorry for disappearing, had a lot to do.....

There's a new commit here, this should solve your issue about the encrypted device freeze.
Please test and report back.
Thanks!

**Comment by kholk** [comment]
P.S.: Remember to update the kernel to the latest HEAD!

**Comment by djselbeck** [comment]
This helps. It applies the whitepoint level after first unlock (after the sim pin) when (for some reason) the panel driver will not freeze on applying the setting

**Comment by kholk** [comment]
PR is ready. Thanks for testing.

MERGEME
